File: baekjoon_all/12000+/boj_12891.java
// Solve 2024-02-01

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_12891 {

    static char[] acgt = { 'A', 'C', 'G', 'T' };

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int s = Integer.parseInt(st.nextToken());
        int p = Integer.parseInt(st.nextToken());
        String dna = br.readLine();
        st = new StringTokenizer(br.readLine(), " ");
        int[] minCnt = new int[26];

        for (int i = 0; i < 4; i++) {
            minCnt[acgt[i] - 'A'] = Integer.parseInt(st.nextToken());
        }

        int begin = 0;
        int end = 0;
        int[] curCnt = new int[26];

        while (end < p) {
            curCnt[dna.charAt(end) - 'A']++;
            end++;
        }

        int ans = 0;

        if (isInRange(minCnt, curCnt)) ans++;

        while (end < s) {
            curCnt[dna.charAt(begin) - 'A']--;
            curCnt[dna.charAt(end) - 'A']++;
            begin++;
            end++;

            if (isInRange(minCnt, curCnt)) ans++;
        }

        System.out.println(ans);
        br.close();
    }

    static boolean isInRange(int[] minCnt, int[] curCnt) {
        for (char x : acgt) {
            if (curCnt[x - 'A'] < minCnt[x - 'A']) {
                return false;
            }
        }

        return true;
    }

}

File: baekjoon_all/12000+/boj_12891.cpp
// Solve 2024-02-01

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;

int main() {
    FASTIO;

    int s, p;
    string dna;
    cin >> s >> p >> dna;

    string acgt = "ACGT";
    int min_cnt[26];

    for (int i = 0; i < 4; i++) {
        cin >> min_cnt[acgt[i] - 'A'];
    }

    int begin = 0;
    int end = 0;
    int cur_cnt[26] = { 0 };

    while (end < p) {
        cur_cnt[dna[end] - 'A']++;
        end++;
    }

    int ans = 0;
    bool flag = true;

    for (int i = 0; i < 4; i++) {
        if (cur_cnt[acgt[i] - 'A'] < min_cnt[acgt[i] - 'A']) {
            flag = false;
            break;
        }
    }

    if (flag) ans++;

    while (end < s) {
        cur_cnt[dna[begin] - 'A']--;
        cur_cnt[dna[end] - 'A']++;
        begin++;
        end++;

        flag = true;

        for (int i = 0; i < 4; i++) {
            if (cur_cnt[acgt[i] - 'A'] < min_cnt[acgt[i] - 'A']) {
                flag = false;
                break;
            }
        }

        if (flag) ans++;
    }

    cout << ans << '\n';

    return 0;
}
